### ISSUE TITLE
Does not support :is pseudo-class correctly

### ISSUE THREAD

**Issue by setyb** [less/less.js#3622]
Here are 2 examples of valid code using the `:is` pseudo-class that are not handled by LESS correctly.  Can you please allow valid code like this to work with LESS?

Tested on: https://lesscss.org/less-preview/
which erroneously issues the error message: `Missing closing ')'` for both examples.

```
@-moz-document domain("example.com") {
	:is([foo], [bar] /* :is(a, b) */, [baz]) {
		color: red; }
}
```

```
@-moz-document domain("example.com") {
		:is(
			[a1],
			[a2]
		):is(
			[a3],
			[a4]:not(
				:is([a5],[a6])
			):is(
				[a7]:not(:is([a8],[a9])),
				[b1]:not(:is([b2],[b3]))
			):is(
				[b4],[b5]
			),
			[b6],
			[b7]
		) {
			color: red; }
	
}
```

See also: https://github.com/openstyles/stylus/issues/1253

**Comment by skuIIs** [comment]
The bug includes comments or variables inside of the :is() selectors

Example with the `Missing closing ')'` error

```
each(@color, {
  :is(.color-@{key},
  .focus-color-@{key}:focus,
  .hover-color-@{key}:hover) {
    color: @value !important;
  }
});
```

**Comment by iChenLei** [comment]
### TL;DR
Less not support write expression(and comment which contain &()@ character ) in pseudo-class function currently.

----------

### functional css pseudo-class from [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-classes):

1. [:dir()](https://developer.mozilla.org/en-US/docs/Web/CSS/:dir) 🧪
2. [:has()](https://developer.mozilla.org/en-US/docs/Web/CSS/:has) 🧪
3. [:host()](https://developer.mozilla.org/en-US/docs/Web/CSS/:host()) 🧪
4. [:host-context()](https://developer.mozilla.org/en-US/docs/Web/CSS/:host-context()) 🧪
5. [:is()](https://developer.mozilla.org/en-US/docs/Web/CSS/:is) 🧪
6. [:lang()](https://developer.mozilla.org/en-US/docs/Web/CSS/:lang) 
7. [:not()](https://developer.mozilla.org/en-US/docs/Web/CSS/:not)
8. [:nth-child()](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-child)
9. [:nth-col()](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-col)  🧪
10. [:nth-last-child()](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-last-child)
11. [:nth-last-col()](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-last-col) 🧪
12. [:nth-last-of-type()](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-last-of-type) 
13. [:nth-of-type()](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-of-type)
14. [:state()](https://developer.mozilla.org/en-US/docs/Web/CSS/:state) 🧪
15. [:where()](https://developer.mozilla.org/en-US/docs/Web/CSS/:where) 🧪

> 🧪   means experimental feature

#### A minimal reproducation code snippet for `skuIIs` and `setyb`'s issue.
```less
@selector: p;
:is(@selector) {
  color: red;
}
```` 
```diff
- Missing closing ')'
```

#### Why these code work?
```less
@color: red;
:is(/* :is */ .color-lei, .color) {
  color: @color;
}
```
compile to
```less
:is(/* :is */ .color-lei, .color) {
  color: red;
}
```
Let's try to debug it ->

![2021-07-06 20 32 22](https://user-images.githubusercontent.com/14012511/124694942-ce650900-df14-11eb-8280-755643a2ea76.png)

`:is(/* :is */ .color-lei, .color)` entity was treated like `selectors`, `:is` and `(/* :is */ .color-lei, .color)` are two elements of `selectors`.
> From a source code(master branch) perspective, we can look the detail of less parse process.

https://github.com/less/less.js/blob/36bb6cef7c23034e9fdd4b64bd123257ac73a445/packages/less/src/less/parser/parser.js#L1269-L1282

`parserInput.$re(/^\([^&()@]+\)/)` match `(balabala....)` as a element, but it can't contain `&` `(` `)` and `@` otherwise parse fail.

![2021-07-07 11 30 47](https://user-images.githubusercontent.com/14012511/124695968-c908be00-df16-11eb-9f27-0dea7bcaf308.png)

That's why when you add a comment node which contain `(` will cause parse fail.
```less
// error
:is([foo], [bar] /* :is(a, b) */, [baz]) {
    color: red; 
}
// ok
:is([foo], [bar] /* :is */, [baz]) {
    color: red; 
}
// error
@selector: p;
:is(@selector) {
    color: red; 
}
// ok
:is(p) {
    color: red; 
}
```
### Conclusion
I think this is a `feature request` and not `bug fix`, terminal error message ouput misleading us. The question is `Less not support write expression(and comment which contain &()@ character ) in pseudo-class function currently`.

/cc @matthew-dean Sir, How do you thought about this ?

**Comment by chipx86** [comment]
The analysis here is great.

As these new selectors have matured, we've started to investigate moving over to them incrementally, but got bit by this bug. I'm curious if there are any plans to rework the affected parts of the parser that are getting tripped up here.

If not, is there at least a favored approach to resolving it in case we decide to attempt a fix? A cursory, naive read through the parser code suggests that these newer selectors probably would need to be explicitly handled, and that logic could be less strict than the general selector logic here.

I wouldn't mind playing around with a potential fix if the team here is open to it, but I don't want to go down the wrong path in a fix, or duplicate existing work.

**Comment by puckowski** [comment]
I have fixes for the following syntax that passes all tests:

```css
@-moz-document domain("example.com") {
	:is([foo], [bar] /* :is(a, b) */, [baz]) {
		color: red; }
}
```

```css
@-moz-document domain("example.com") {
		:is(
			[a1],
			[a2]
		):is(
			[a3],
			[a4]:not(
				:is([a5],[a6])
			):is(
				[a7]:not(:is([a8],[a9])),
				[b1]:not(:is([b2],[b3]))
			):is(
				[b4],[b5]
			),
			[b6],
			[b7]
		) {
			color: red; }
	
}
```

```css
each(@color, {
  :is(.color-@{key},
  .focus-color-@{key}:focus,
  .hover-color-@{key}:hover) {
    color: @value !important;
  }
});
```

Will try to send a PR this weekend.

**Comment by puckowski** [comment]
PR https://github.com/less/less.js/pull/4298 should resolve some of the parsing issues with syntax raised in this issue.
